### PR TITLE
Add BuiltReduxUiComponent

### PR DESCRIPTION
**Pull request by jacehensley-wf**
## Ultimate problem:
We should have support for the [built_redux](https://github.com/davidmarne/built_redux) library.

## How it was fixed:
- Add `BuiltReduxUiProps` and `BuiltReduxUiComponent` that wire up listeners to the store's stream.

## Testing suggestions:
- Verify tests pass

## Potential areas of regression:
- N/A All new stuff

---

> __FYA:__ @greglittlefield-wf @aaronlademann-wf @clairesarsam-wf @joelleibow-wf
> __FYI:__ @davidmarne-wf

### CODE REVIEW

**Comment by aviary2-wf** [comment]
## Raven
<details>
<summary>Number of Findings: 0</summary>

</details>

File: pubspec.yaml [hunk 1]
```diff
@@ -9,6 +9,8 @@ environment:
 dependencies:
   analyzer: ">=0.30.0 <0.31.0"
   barback: "^0.15.0"
+  built_redux: ^6.1.0
+  built_value: ^4.2.0
```
Comment (greglittlefield-wf) [inline pubspec.yaml]: Is this dependency needed?

File: pubspec.yaml [hunk 2]
```diff
@@ -9,6 +9,8 @@ environment:
 dependencies:
   analyzer: ">=0.30.0 <0.31.0"
   barback: "^0.15.0"
+  built_redux: ^6.1.0
```
Comment (greglittlefield-wf) [inline pubspec.yaml]: This dependency worries me since it's undergoing major version bumps quite frequently. If we include this, that means we have to stay on top of bumping the upper bound to avoid being a source of version lock.

File: pubspec.yaml (hunk 1 above)
Comment (jacehensley-wf) [reply pubspec.yaml]: Hmm I guess it's only needed for tests, will move

File: pubspec.yaml (hunk 2 above)
Comment (jacehensley-wf) [reply pubspec.yaml]: Yeah, that's a good point

**Comment by greglittlefield-wf** [comment]
Is this set of base classes a pattern that's also being used in comments? Trying to get a feel for how mature this approach is, and whether more exploration is needed before committing to this API.

**Comment by jacehensley-wf** [comment]
I believe w_comments it just listening to the streams on mount. Let me double check

**Comment by greglittlefield-wf** [comment]
And to clarify, I'm not against this approach, I just want to make sure it's tried and true before adding it to this repo.

**Comment by jacehensley-wf** [comment]
@greglittlefield-wf You can see their usage [here](https://github.com/Workiva/w_comments/search?utf8=%E2%9C%93&q=%22stream.listen%22&type=)

File: lib/src/component_declaration/redux_component.dart [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+import 'package:meta/meta.dart';
+import 'package:over_react/over_react.dart';
+
+/// Builds on top of [UiProps], adding typed props for [Store] in order to integrate with `built_redux`.
+///
+/// Use with the over_react transformer via the `@Props()` ([annotations.Props]) annotation.
+abstract class ReduxUiProps<StoreT> extends UiProps {
+  String get _storePropKey => '${propKeyNamespace}store';
+
+  /// The prop defined by [StoresT].
```
Comment (aaronlademann-wf) [inline lib/src/component_declaration/redux_component.dart]: s/StoresT/StoreT

File: lib/src/component_declaration/redux_component.dart [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+  ///
+  /// This object should either be an instance of [Store].
```
Comment (aaronlademann-wf) [inline lib/src/component_declaration/redux_component.dart]: either an instance of [Store] ... or ... what else? ;)

File: test/over_react/component_declaration/redux_component_test/test_reducer.g.dart [hunk 5]
```diff
@@ -0,0 +1,222 @@
+// GENERATED CODE - DO NOT MODIFY BY HAND
```
Comment (aaronlademann-wf) [inline test/over_react/component_declaration/redux_component_test/test_reducer.g.dart]: @jacehensley-wf is it necessary to commit generated files?

File: tool/build.dart [hunk 6]
```diff
@@ -0,0 +1,19 @@
+import 'dart:async';
+
+import 'package:build_runner/build_runner.dart';
+import 'package:built_value_generator/built_value_generator.dart';
+import 'package:source_gen/source_gen.dart';
+import 'package:built_redux/generator.dart';
+
+/// Build the generated files in the built_value chat example.
```
Comment (aaronlademann-wf) [inline tool/build.dart]: What chat example?

File: lib/src/component_declaration/redux_component.dart [hunk 7]
```diff
@@ -0,0 +1,163 @@
+// Copyright 2016 Workiva Inc.
```
[the remark this answers is not in the document]
Comment (aaronlademann-wf) [reply lib/src/component_declaration/redux_component.dart]: Same goes for all files added in this PR

File: pubspec.yaml (hunk 2 above)
Comment (johnbland-wf) [reply pubspec.yaml]: It only looks that way because, per @davidmarne-wf, because he went 1.0 too early and broke some things so he had to rev for semver purposes.

File: test/over_react/component_declaration/redux_component_test/test_reducer.g.dart (hunk 5 above)
Comment (johnbland-wf) [reply test/over_react/component_declaration/redux_component_test/test_reducer.g.dart]: If you don't commit, build it into your build process to build before tests run.

File: test/over_react/component_declaration/redux_component_test/default.dart [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+@Component()
+class TestDefaultComponent extends ReduxUiComponent {
+  int numberOfRedraws = 0;
+
+  @override
+  render() => Dom.div()();
+
+  @override
+  void setState(_, [callback()]) {
+    print('here');
```
Comment (johnbland-wf) [inline test/over_react/component_declaration/redux_component_test/default.dart]: log

File: lib/src/component_declaration/redux_component.dart [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+  }
+
+  @mustCallSuper
+  @override
+  void componentWillReceiveProps(Map nextProps) {
+    super.componentWillReceiveProps(nextProps);
+    // Unbind store subs so they can be re-bound in componentDidUpdate
+    // once the new props are available, ensuring the values returned by [redrawOn]
+    // are not outdated.
+    _tearDownSubs();
```
Comment (johnbland-wf) [inline lib/src/component_declaration/redux_component.dart]: If the store prop is the same, we don't need new subs, right?

**Comment by johnbland-wf** [comment]
@greglittlefield-wf, the preferred use in Redux.js is to use a HOC called a provider and connect the data primitive components. In Dart, we don't have that until react-dart supports context.

We're starting to use this in grc and we're looking at a wrapper component vs a base component. The typical flow, per David, is to have a wrapper component handle the subscription and wire up the child components from there. This PR would be used as base class(es) we extend for our wrapper component.

**Comment by johnbland-wf** [comment]
Oh and you really should bring over the pure component in Comments. If we're using built-redux, a `PureReduxComponent` should exist for everyone.

File: test/over_react/component_declaration/redux_component_test/test_reducer.g.dart (hunk 5 above)
Comment (jacehensley-wf) [reply test/over_react/component_declaration/redux_component_test/test_reducer.g.dart]: Yeah, I can just build it during CI, but while deving you'll have analysis warnings till you generate

File: lib/src/component_declaration/redux_component.dart (hunk 9 above)
Comment (jacehensley-wf) [reply lib/src/component_declaration/redux_component.dart]: It's not worth checking that, it doesn't take much to teardown and re-bind the subs

**Comment by jacehensley-wf** [comment]
@johnbland-wf We can't do the HOC until we add support for React's context to react-dart.

For the `PureReduxComponent` I could see that being useful but I'd rather just get this in for now, and adding something like that later.

File: pubspec.yaml (hunk 2 above)
Comment (davidmarne-wf) [reply pubspec.yaml]: Yeah, I understand your hesitation for sure, but I really do believe it finally settled down. I'm very happy with where it stands right now.

Plus at this point it is a dependency for CDS, w_comments, review_bar, and soon to be graph_app (and many other repos via transitivity from w_comments) so any major bumps are already going to have to be dealt with anyways. 

However, if we want I could easy make another repo that just has this component.

File: lib/src/component_declaration/redux_component.dart [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+
+  @mustCallSuper
+  @override
+  void componentWillUnmount() {
+    super.componentWillUnmount();
+    _tearDownSubs();
+  }
+}
+
+abstract class _ReduxComponentMixin<T extends ReduxUiProps> implements UiComponent<T> {
```
Comment (davidmarne-wf) [inline lib/src/component_declaration/redux_component.dart]: I think it would make a lot of sense to have a connect function here. The default implementation could be something like

```dart
Substate connect(State state) => state;
```

But then that could be overridden to only redraw when specific pieces of your state change

```dart
@override
Substate connect(State state) => state.myComponentWantsToRedrawOn;
```

then redrawOn would be something like
```dart
List<Stream> redrawOn() {
     if (props.store is Store) {
       return <Stream>[props.store.nextSubstate(connect)];
     } else {
       return <Stream>[];
     }
  }
```

Comment (jacehensley-wf) [reply lib/src/component_declaration/redux_component.dart]: With that how would someone listen to multiple stores?

Comment (johnbland-wf) [reply lib/src/component_declaration/redux_component.dart]: There is no concept of multiple stores. It is one store in redux.

**Comment by johnbland-wf** [comment]
@jacehensley-wf yes (re: react context). I mentioned that in hopes @greglittlefield-wf would find it in his heart to push that forward. ;)

Agreed on `PureReduxComponent` being separate. Just getting the list out there. ;)

**Comment by greglittlefield-wf** [comment]
My main concern was marching forward with this without thoroughly planning this feature, but it seems like there's a consensus that this is what we want 👍.

File: lib/src/component_declaration/redux_component.dart [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+  StoreT get store => props[_storePropKey];
+  set store(StoreT value) => props[_storePropKey] = value;
+}
+
+/// Builds on top of [UiComponent], adding `built_redux` integration.
+///
+/// * Redux components are responsible for rendering application views and turning
+///   user interactions and events into [Action]s.
+/// * Redux components can use data from one or many [Store] instances to define
+///   the resulting component.
```
Comment (evanweible-wf) [inline lib/src/component_declaration/redux_component.dart]: Just one store, right?

Comment (johnbland-wf) [reply lib/src/component_declaration/redux_component.dart]: Yup

File: lib/src/component_declaration/redux_component.dart (hunk 9 above)
Comment (johnbland-wf) [reply lib/src/component_declaration/redux_component.dart]: I'm curious why it isn't worth it. It seems like you'd want to not force something that is not necessary.

Comment (greglittlefield-wf) [reply lib/src/component_declaration/redux_component.dart]: Tearing down and rebinding subscriptions is cheap, I don't think it's necessary to expend effort to avoid it.

Also, we we'd have to compare the return value of `redrawOn` and not just `props.store`, and we don't have access to that updated return value until `componentDidUpdate` (since that function can rely on values within `props`). So, we'd have to store the old values in here and then check them and null them out in componentDidUpdate.

In this case, unbinding/rebinding is more straightforward, and again, should cheap enough to be worth it.

File: lib/src/component_declaration/redux_component.dart (hunk 10 above)
Comment (jacehensley-wf) [reply lib/src/component_declaration/redux_component.dart]: True, what about multiple values on the single store?

Comment (johnbland-wf) [reply lib/src/component_declaration/redux_component.dart]: Agreed. We'll want a substate on multiple values.

Comment (davidmarne-wf) [reply lib/src/component_declaration/redux_component.dart]: So one thing you can do is map all the values into a new built_value. For example say by store's state object is:
```dart
abstract class AppState implements Built<AppState, AppStateBuilder> {
  int get foo;
  int get bar;
  int get baz;
}
```

But my component only wants foo and bar. I can still only using a single nextSubstate by defining a new model that contains the subset of state my component cares about and having my connect function return that. So i would define

```dart
abstract class DataComponentWants implements Built<DataComponentWants, DataComponentWantsBuilder> {
  int get foo;
  int get bar;
}
```

and have connect do:
```dart
DataComponentWants connect(AppState state) => new DataComponentWants(
  foo: state.foo,
  bar: state.bar,
);
```

Comment (davidmarne-wf) [reply lib/src/component_declaration/redux_component.dart]: Since DataComponentWants is also a built, it should be comparable with the last result of connect, then we can limit redraws by only redrawing if the last result == new result

File: lib/src/component_declaration/redux_component.dart (hunk 9 above)
Comment (davidmarne-wf) [reply lib/src/component_declaration/redux_component.dart]: I'm confused about this, tracking _areStoreSubsBound and having all these teardown and setups it seems overly complicated to me. 
The reference to your store should never change, it seems to me that you would only setup subs on componentWillMount and remove them on componentWillUnmount.

**Comment by davidmarne-wf** [comment]
I threw up a proposal on how I could see this working here:
https://github.com/davidmarne-wf/over_react/pull/1/files

The generics get a bit verbose, but it is necessary to get legit type safety.

**Comment by jacehensley-wf** [comment]
@davidmarne-wf How does this look?

File: lib/src/component_declaration/redux_component.dart [hunk 12]
```diff
@@ -47,117 +51,68 @@ abstract class ReduxUiProps<StoreT> extends UiProps {
...
-
-    _storeSubs = [];
-
-    for (var stream in redrawOn()) {
-      _storeSubs.add(stream.listen((_) => redraw()));
-    }
+    _storeSub = props.store.nextSubstate(connect).listen((Substate s) {
+      _connectedProps = s;
+      redraw();
+    });
```
Comment (johnbland-wf) [inline lib/src/component_declaration/redux_component.dart]: I think the only problem with this is we can only listen to one substate so we'd be stuck if we wanted `state.count` and `state.todos`.

File: lib/src/component_declaration/redux_component.dart [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+  ///         baz: baz,
+  ///       );
+  ///       MyState._();
+  ///
+  ///       int get foo;
+  ///       int get bar;
+  ///       int get baz;
+  ///     }
+  ///
+  ///     abstract class DataCoponentCaresAbout implements Built<DataCoponentCaresAbout, DataCoponentCaresAboutBuilder> {
```
Comment (johnbland-wf) [inline lib/src/component_declaration/redux_component.dart]: `DataCoponent` -> `DataComponent`

File: lib/src/component_declaration/redux_component.dart [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+  ///       render() {
+  ///         return Dom.dov()(
+  ///           connectedState.foo,
+  ///           connectedState.bar,
+  ///         );
+  ///       }
+  ///     }
+  ///
+  /// Related: [connectedState]
+  Substate connect(V state);
```
Comment (davidmarne-wf) [inline lib/src/component_declaration/redux_component.dart]: I think it would be worth adding to the docstring that Substate must be a comparable object to avoid unnecessary redraws. Add a recommendation that Substate should be either primitive, built_value, built_collection, or other type of object that overrides the == operator.

Comment (davidmarne-wf) [reply lib/src/component_declaration/redux_component.dart]: as well as a similar blurb to the ReduxUiComponent docstring

File: lib/src/component_declaration/redux_component.dart (hunk 12 above)
Comment (johnbland-wf) [reply lib/src/component_declaration/redux_component.dart]: Thoughts, @davidmarne-wf?

Comment (jacehensley-wf) [reply lib/src/component_declaration/redux_component.dart]: No, your substate would be a built_value object like shown in the doc comment https://github.com/Workiva/over_react/pull/118/files/20dc99def5af7766e5045dbcc909ad3bb39db8f1#diff-cd6ab537d255abbf1ec247df0369120eR119

Comment (davidmarne-wf) [reply lib/src/component_declaration/redux_component.dart]: i am of the opinion you define a new built_value and map state to that. so say your state is
```dart
abstract class State extends Built<State, StateBuilder> {
  int get count;
  BuiltList<Todo> todos;
  Map<String, String> options;
 ...
}
```
you define a model
```dart
abstract class SubstateMyComponentWants extends Built<SubstateMyComponentWants, SubstateMyComponentWantsBuilder> {
  int get count;
  BuiltList<Todo> todos;
 ...
}
```

then your connect is 
```dart
@override
SubstateMyComponentWants connect(State s) => new SubstateMyComponentWants((builder) => builder
  ..count = state.count
  ..todos = state.todos);
```

Comment (davidmarne-wf) [reply lib/src/component_declaration/redux_component.dart]: I'ld be open to suggestions for multiple connectors though. I imagine it would look something like:
```dart
typedef Substate StateMapper<State, Substate>(State s);

Iterable<StateMapper<State, dynamic>> get connectors;

...

void _setUpSub() {
    _storeSubs.addAll(connectors.map((connect) => props.store.nextSubstate(connect).listen((_) {
      redraw();
    }));
  }
```

Comment (davidmarne-wf) [reply lib/src/component_declaration/redux_component.dart]: then you wouldn't have connectedState, you would prob just expose a store.state getter and let the renderers pull data from that.

Comment (davidmarne-wf) [reply lib/src/component_declaration/redux_component.dart]: or even simpler you could just have 

```dart
bool shouldUpdateOnReduxStateChange(State s);
```

that the consumer implements.

then if i wanna update on todos or count changes i do:

```dart
@override
bool shouldUpdateOnReduxStateChange(State prev, State next) =>  prev.todos != next.todos || prev.count != next.count
```

The upside to keeping how it is now, where you load everything into a new built_value, is this comparison is generated for the model

Comment (jacehensley-wf) [reply lib/src/component_declaration/redux_component.dart]: I like the first option, with using built_value objects to contain multiple state values

Comment (johnbland-wf) [reply lib/src/component_declaration/redux_component.dart]: That's fine.

**Comment by jacehensley-wf** [comment]
@davidmarne-wf Added that doc comment :)

**Comment by johnbland-wf** [comment]
Maybe this should be renamed `BuiltReduxUiComponent` since it is specific to built_redux only and not redux-dart.

**Comment by jacehensley-wf** [comment]
That's too wordy IMO

**Comment by johnbland-wf** [comment]
I agree. It'll be confusing to someone trying to use Redux directly when this won't work for them.

File: lib/src/component_declaration/redux_component.dart [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+abstract class ReduxUiProps<V extends Built<V, B>, B extends Builder<V, B>,
+    A extends ReduxActions> extends UiProps {
+  String get _storePropKey => '${propKeyNamespace}store';
+
+  /// The [Store] prop defined by [V], [B], and [A].
+  Store<V, B, A> get store => props[_storePropKey];
+  set store(Store<V, B, A> value) => props[_storePropKey] = value;
+
+  /// The [ReduxActions] prop defined by [A].
+  A get actions => store.actions;
```
Comment (greglittlefield-wf) [inline lib/src/component_declaration/redux_component.dart]: Can we add test coverage for this?

File: lib/src/component_declaration/redux_component.dart [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+  void componentWillMount() {
+    super.componentWillMount();
+    _setUpSub();
+  }
+
+  @mustCallSuper
+  @override
+  void componentWillReceiveProps(Map nextProps) {
+    super.componentWillReceiveProps(nextProps);
+    _tearDownSub();
```
Comment (greglittlefield-wf) [inline lib/src/component_declaration/redux_component.dart]: Looks like this method is missing test coverage

File: lib/src/component_declaration/redux_component.dart [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+  }
+
+  @mustCallSuper
+  @override
+  void componentWillUpdate(Map nextProps, Map nextState) {
+    if (_storeSub == null) _setUpSub(nextProps);
```
Comment (greglittlefield-wf) [inline lib/src/component_declaration/redux_component.dart]: Looks like this method is missing test coverage

File: lib/src/component_declaration/redux_component.dart [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+
+abstract class _ReduxComponentMixin<
+    V extends Built<V, B>,
+    B extends Builder<V, B>,
+    A extends ReduxActions,
+    T extends ReduxUiProps<V, B, A>,
+    Substate> implements UiComponent<T> {
+  Substate _connectedState;
+
+  /// The substate values of the redux store that this component component subscribes to.
```
Comment (greglittlefield-wf) [inline lib/src/component_declaration/redux_component.dart]: typo: `component component`

File: lib/src/component_declaration/redux_component.dart [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+  ///
+  /// It is reccommened to use this instead of `props.store.state`,
```
Comment (greglittlefield-wf) [inline lib/src/component_declaration/redux_component.dart]: #nit comma should be a semicolon

File: lib/src/component_declaration/redux_component.dart [hunk 20]
```diff
@@ -0,0 +1,171 @@
...
+  ///       int get foo;
+  ///       int get bar;
+  ///     }
+  ///
+  ///     MyComponent extends ReduxUiComponent<MyState, MyStateBuilder, MyActions, MyProps, DataComponentCaresAbout> {
+  ///       @override
+  ///       DataComponentCaresAbout connect(state) => new DataComponentCaresAbout(foo: state.foo, bar: state.bar);
+  ///
+  ///       render() {
+  ///         return Dom.dov()(
```
Comment (greglittlefield-wf) [inline lib/src/component_declaration/redux_component.dart]: typo: `dov`

File: lib/src/component_declaration/redux_component.dart [hunk 21]
```diff
@@ -0,0 +1,171 @@
...
+  ///           connectedState.foo,
+  ///           connectedState.bar,
+  ///         );
+  ///       }
+  ///     }
+  ///
+  /// __Note:__ [Substate] must be a comparable object to avoid unnecessary redraws,
```
Comment (greglittlefield-wf) [inline lib/src/component_declaration/redux_component.dart]: #nit comma should be semicolon

File: lib/src/component_declaration/redux_component.dart [hunk 22]
```diff
@@ -0,0 +1,171 @@
...
+  /// it is reccomended that [Substate] either be a primitive, built_value, build_collection or an Object
```
Comment (greglittlefield-wf) [inline lib/src/component_declaration/redux_component.dart]: Typo? Is `build_collection` supposed to be `built_collection`?

File: lib/src/component_declaration/redux_component.dart [hunk 23]
```diff
@@ -0,0 +1,171 @@
...
+  /// that overrides `==`.
```
Comment (greglittlefield-wf) [inline lib/src/component_declaration/redux_component.dart]: It'd probably be a good idea to mention overriding `hashCode` as well, since it's not immediately obvious to all developers that you need to override both and not just one.

File: lib/src/component_declaration/redux_component.dart [hunk 24]
```diff
@@ -0,0 +1,171 @@
...
+
+  /// The substate values of the redux store that this component component subscribes to.
+  ///
+  /// It is reccommened to use this instead of `props.store.state`,
+  /// if you need to access other state values from the store then they should be connected with [connect].
+  ///
+  /// Related: [connect]
+  Substate get connectedState => _connectedState;
+
+  /// Subscribe to changes to the values from the redux store that this component cares about.
```
Comment (greglittlefield-wf) [inline lib/src/component_declaration/redux_component.dart]: #nit This doesn't follow the Dart style guide convention around the first line of function doc comments.

Should be something like:
> Returns a subset of the values derived from the redux store that this component cares about.

File: lib/src/component_declaration/redux_component.dart (hunk 17 above)
Comment (greglittlefield-wf) [inline lib/src/component_declaration/redux_component.dart]: Might be good to add a code comment here that explains when this is null (i.e., for component props updates and not just any rerender)

File: test/over_react/component_declaration/redux_component_test.dart [hunk 25]
```diff
@@ -0,0 +1,124 @@
...
+      var store = new Store<BaseState, BaseStateBuilder, BaseActions>(
+        baseReducerBuilder.build(),
+        baseState,
+        baseActions,
+      );
+      var renderedInstance = render(TestDefault()..store = store);
+      TestDefaultComponent component = getDartComponent(renderedInstance);
+
+      store.actions.trigger1();
+      await new Future.delayed(Duration.ZERO);
```
Comment (greglittlefield-wf) [inline test/over_react/component_declaration/redux_component_test.dart]: Is this necessary now that actions dispatch synchronously? Should we set the lower bound to the major version that changes that?

File: test/over_react/component_declaration/redux_component_test.dart [hunk 26]
```diff
@@ -0,0 +1,124 @@
...
+      unmount(renderedInstance);
+
+      store.actions.trigger1();
+      await new Future.delayed(Duration.ZERO);
+
+      expect(component.numberOfRedraws, 1,
+          reason: 'component should no longer be listening after unmount');
+    });
+
+    test('subscribes to any state changes subscribed to in connect', () async {
```
Comment (greglittlefield-wf) [inline test/over_react/component_declaration/redux_component_test.dart]: This description would be clearer and more indicative of what's being tested if it read "only the state changes subscribed to in connect".

File: test/over_react/component_declaration/redux_component_test.dart [hunk 27]
```diff
@@ -0,0 +1,124 @@
...
+      );
+      var renderedInstance = render((TestDefault()..store = store)());
+      TestDefaultComponent component = getDartComponent(renderedInstance);
+
+      store.actions.trigger1();
+      await new Future.delayed(Duration.ZERO);
+      expect(component.numberOfRedraws, 1);
+
+      component.props = TestDefault()..store = updatedStore;
+      component.redraw();
```
Comment (greglittlefield-wf) [inline test/over_react/component_declaration/redux_component_test.dart]: This should be actually rerendering the component and not just simulating a props update.

File: test/over_react/component_declaration/redux_component_test.dart (hunk 25 above)
Comment (jacehensley-wf) [reply test/over_react/component_declaration/redux_component_test.dart]: I wanted to leave it open to both version to not be blocking, I don't think w_comments is on that major version yet

File: lib/src/component_declaration/redux_component.dart (hunk 15 above)
Comment (jacehensley-wf) [reply lib/src/component_declaration/redux_component.dart]: Done

File: lib/src/component_declaration/redux_component.dart (hunk 16 above)
Comment (jacehensley-wf) [reply lib/src/component_declaration/redux_component.dart]: Done

**Comment by johnbland-wf** [comment]
One more attempt for a rename here. ;)

File: test/over_react/component_declaration/redux_component_test.dart (hunk 25 above)
Comment (greglittlefield-wf) [reply test/over_react/component_declaration/redux_component_test.dart]: 👍 

If you end up making changes in this PR again, perhaps a todo comment would be good to remove these if we ever bump. Waiting extra long shouldn't hurt, but it would be nice to verify that the things we expect to be synchronous are indeed synchronous.

**Comment by greglittlefield-wf** [comment]
+1

**Comment by brianbolton-wf** [comment]
FYI @christianplazas-wf @garywhitlock-wf

**Comment by aaronlademann-wf** [comment]
+1 refresh

**Comment by greglittlefield-wf** [comment]
+1

**Comment by jayudey-wf** [comment]
QA +1

- [ ] Testing instruction
- [ ] Dev +1's
- [ ] Dev/QA +10 with detail of what was tested
  - passing build with tests exercising new functionality
- [ ] Security review (if required)
- [ ] Unit tests created/updated
- [ ] All unit tests pass
- [ ] Rosie has run and reports properly the release the ticket will be included in
- [ ] Areas of regression noted

Merging into master.
